# 2장 1강: t-검정의 이론과 가정 — 실습문제

## 실습 목표

- 평균 차이와 표준오차를 이용해 t통계량을 직접 계산할 수 있다.
- 단일표본 t검정으로 하나의 표본평균과 기준값을 비교할 수 있다.
- 독립성·정규성·등분산성을 확인한 뒤 독립표본 t검정을 수행할 수 있다.
- 독립표본과 대응표본 상황을 구분하고 적절한 함수를 선택할 수 있다.
- p-value를 유의수준과 비교하여 검정 결과를 올바르게 해석할 수 있다.

## 실습 환경 / 데이터

- Python
- NumPy, pandas
- scipy.stats
- `ames_housing.csv`

| 컬럼 | 의미 |
|---|---|
| `SalePrice` | 주택 판매가격 |
| `KitchenQual` | 주방 품질 |

> 모든 검정은 양측검정이며 유의수준 `α = 0.05`를 사용합니다.  
> Ames 표본 추출에는 지정된 `random_state`를 사용해 결과를 재현합니다.

## 실습 준비

1. 필요한 라이브러리를 불러오세요.
2. Ames Housing 데이터를 `df`에 불러오세요.
3. 데이터 크기, 전체 결측치 수, 컬럼명과 상위 5개 행을 확인하세요.

In [1]:
# 실습 준비 코드를 작성하세요.
from pathlib import Path
import numpy as np
import pandas as pd
from scipy import stats

df = pd.read_csv("ames_housing.csv")
df.head()

,SalePrice,GrLivArea,LotArea,OverallQual,KitchenQual,CentralAir,HeatingQC,PavedDrive,Neighborhood,YearBuilt
0,208500,1710,8450,7,Gd,Y,Ex,Y,CollgCr,2003
1,181500,1262,9600,6,TA,Y,Ex,Y,Veenker,1976
2,223500,1786,11250,7,Gd,Y,Ex,Y,CollgCr,2001
3,140000,1717,9550,7,Gd,Y,Gd,Y,Crawfor,1915
4,250000,2198,14260,8,Gd,Y,Ex,Y,NoRidge,2000


---

## 필수 1. t통계량 직접 계산과 단일표본 t검정

### 문제 1-1. 평균 판매가격은 180,000달러와 다른가?

#### 문제 설명

Ames 주택 30개를 표본으로 추출하여 평균 판매가격이 기준값 180,000달러와 다른지 확인합니다. 먼저 t통계량을 직접 계산한 뒤 `ttest_1samp()` 결과와 비교하세요.

#### 요구사항

1. `SalePrice`에서 `n=30`, `random_state=2`로 표본을 추출하여 `sale_sample`에 저장하세요.
2. 표본 수, 표본평균, 표본표준편차를 출력하세요.
3. `stats.sem()`을 이용해 표준오차를 계산하세요.
4. 다음 식으로 t통계량을 직접 계산하세요.  
   `t = (표본평균 - 기준값) / 표준오차`
5. Shapiro-Wilk 검정으로 표본의 정규성을 확인하세요.
6. 다음 가설을 작성하세요.
   - H₀: 모집단 평균 판매가격은 180,000달러이다.
   - H₁: 모집단 평균 판매가격은 180,000달러가 아니다.
7. `stats.ttest_1samp()`로 단일표본 t검정을 수행하세요.
8. 직접 계산한 t통계량과 함수가 반환한 t통계량을 비교하세요.
9. p-value를 이용해 귀무가설 기각 여부를 판단하세요.

#### 해석 질문

**Q1.** t통계량은 평균 차이와 표준오차를 어떻게 이용한 값인가요?  
**Q2.** 같은 평균 차이라면 표준오차가 작아질수록 t통계량의 절댓값은 어떻게 변하나요?  
**Q3.** 직접 계산한 t통계량과 `ttest_1samp()`의 t통계량은 일치하나요?  
**Q4.** 검정 결과 평균 판매가격이 180,000달러와 다르다고 판단할 수 있나요?

#### 제출 결과

- 기술통계량과 정규성 결과
- 직접 계산한 표준오차와 t통계량
- 단일표본 t검정 결과
- 귀무가설 판단과 해석
- Q1~Q4 답변

In [2]:
# 필수 1 코드를 작성하세요.
alpha = 0.05
standard_price = 180000

sale_sample = df["SalePrice"].dropna().sample(n=30, random_state=2)

sample_mean = sale_sample.mean()
sample_std = sale_sample.std(ddof=1)
standard_error = stats.sem(sale_sample)
manual_t = (sample_mean - standard_price) / standard_error
shapiro_result = stats.shapiro(sale_sample)

# H0: 모집단 평균 판매가격은 180,000달러이다.
# H1: 모집단 평균 판매가격은 180,000달러가 아니다.
t_result = stats.ttest_1samp(sale_sample, popmean=standard_price)

print(f"표본 수: {len(sale_sample)}")
print(f"표본평균: {sample_mean:,.2f}달러")
print(f"표본표준편차: {sample_std:,.2f}")
print(f"표준오차: {standard_error:,.4f}")
print(f"정규성 p-value: {shapiro_result.pvalue:.4f}")
print(f"직접 계산한 t통계량: {manual_t:.4f}")
print(f"함수가 계산한 t통계량: {t_result.statistic:.4f}")
print(f"p-value: {t_result.pvalue:.4f}")
print("t통계량 일치 여부:", np.isclose(manual_t, t_result.statistic))
print("최종 판단:", "귀무가설 기각" if t_result.pvalue <= alpha else "귀무가설 기각 못 함")


표본 수: 30
표본평균: 233,323.80달러
표본표준편차: 86,013.01
표준오차: 15,703.7556
정규성 p-value: 0.2133
직접 계산한 t통계량: 3.3956
함수가 계산한 t통계량: 3.3956
p-value: 0.0020
t통계량 일치 여부: True
최종 판단: 귀무가설 기각


### 필수 1 답변 작성란

**Q1.** t통계량은 평균 차이와 표준오차를 어떻게 이용한 값인가요?  
-> 관측된 평균과 기준값의 차이를 (그 차이의 불확실성을 의미하는) 표준오차로 나눈 값

**Q2.** 같은 평균 차이라면 표준오차가 작아질수록 t통계량의 절댓값은 어떻게 변하나요?  
->  분모(표준오차)가 작아지므로 t통계량의 절대값은 커진다.

**Q3.** 직접 계산한 t통계량과 `ttest_1samp()`의 t통계량은 일치하나요?  
-> 일치한다. 두 방법 모두 약 3.3956으로 계산됨.

**Q4.** 검정 결과 평균 판매가격이 180,000달러와 다르다고 판단할 수 있나요?
-> 그렇다. p-value가 약 0.002로 0.05보다 작아 귀무가설을 기각한다.
-> 따라서 평균판매가격은 180000달러와 다르다.


---

## 필수 2. 독립표본 t검정의 가정 점검과 수행

### 문제 2-1. 주방 품질 `Gd`와 `TA` 집단의 평균 판매가격 비교

#### 문제 설명

주방 품질이 `Gd`인 주택과 `TA`인 주택에서 각각 20개를 추출해 평균 판매가격을 비교합니다. 두 집단은 서로 다른 주택으로 구성되어 있습니다.

#### 요구사항

1. `KitchenQual == "Gd"`와 `KitchenQual == "TA"` 집단의 `SalePrice`에서 각각 `n=20`, `random_state=42`로 표본을 추출하세요.
2. 두 집단의 표본 수와 평균을 확인하세요.
3. 데이터 수집 구조를 근거로 두 집단의 독립성을 설명하세요.
4. 각 집단에 Shapiro-Wilk 정규성 검정을 수행하세요.
5. 두 집단에 Levene 등분산 검정을 수행하세요.
6. 가정 점검 결과에 따라 `equal_var=True` 또는 `equal_var=False`를 결정하세요.
7. `stats.ttest_ind()`로 독립표본 t검정을 수행하세요.
8. t통계량과 p-value를 출력하고 두 집단 평균 차이를 해석하세요.

#### 해석 질문

**Q1.** 두 집단이 독립표본인 이유는 무엇인가요?  
**Q2.** 독립성은 별도의 p-value로 확인할 수 있나요?  
**Q3.** 정규성과 등분산성 가정은 각각 충족되나요?  
**Q4.** `equal_var`에는 어떤 값을 사용해야 하나요?  
**Q5.** 두 집단의 평균 판매가격에는 통계적으로 유의한 차이가 있나요?

#### 제출 결과

- 집단별 표본 수와 평균
- 독립성 설명
- 정규성 및 등분산성 검정 결과
- `equal_var` 선택 근거
- 독립표본 t검정 결과와 해석
- Q1~Q5 답변

In [3]:
# 필수 2 코드를 작성하세요.
alpha = 0.05

group_gd = (
    df.loc[df["KitchenQual"] == "Gd", "SalePrice"]
      .dropna().sample(n=20, random_state=42)
)
group_ta = (
    df.loc[df["KitchenQual"] == "TA", "SalePrice"]
      .dropna().sample(n=20, random_state=42)
)

shapiro_gd = stats.shapiro(group_gd)
shapiro_ta = stats.shapiro(group_ta)
levene_result = stats.levene(group_gd, group_ta)

normal_gd = shapiro_gd.pvalue > alpha
normal_ta = shapiro_ta.pvalue > alpha
equal_variance = levene_result.pvalue > alpha

print(f"Gd: n={len(group_gd)}, 평균={group_gd.mean():,.2f}달러")
print(f"TA: n={len(group_ta)}, 평균={group_ta.mean():,.2f}달러")
print("독립성: 서로 다른 주택으로 구성되어 독립표본")
print(f"Gd 정규성 p-value: {shapiro_gd.pvalue:.4f} → {'정규성 위반 근거 없음' if normal_gd else '정규성 위반'}")
print(f"TA 정규성 p-value: {shapiro_ta.pvalue:.4f} → {'정규성 위반 근거 없음' if normal_ta else '정규성 위반'}")
print(f"등분산성 p-value: {levene_result.pvalue:.4f} → {'등분산성 위반 근거 없음' if equal_variance else '등분산성 위반'}")

t_result = stats.ttest_ind(group_gd, group_ta, equal_var=equal_variance)

print(f"equal_var: {equal_variance}")
print(f"t통계량: {t_result.statistic:.4f}")
print(f"p-value: {t_result.pvalue:.4f}")
print("최종 판단:", "평균 판매가격에 유의한 차이가 있음" if t_result.pvalue <= alpha else "평균 판매가격 차이를 확인하지 못함")

Gd: n=20, 평균=191,916.60달러
TA: n=20, 평균=137,432.50달러
독립성: 서로 다른 주택으로 구성되어 독립표본
Gd 정규성 p-value: 0.4580 → 정규성 위반 근거 없음
TA 정규성 p-value: 0.2503 → 정규성 위반 근거 없음
등분산성 p-value: 0.0593 → 등분산성 위반 근거 없음
equal_var: True
t통계량: 3.4249
p-value: 0.0015
최종 판단: 평균 판매가격에 유의한 차이가 있음


### 필수 2 답변 작성란

**Q1.** 두 집단이 독립표본인 이유는 무엇인가요?  
-> GD / TA 각 집단이 서로 다른 주택으로 구성되어 하나의 관측값이 다른 관측값과 짝을 이루지 않기 때문이다.

**Q2.** 독립성은 별도의 p-value로 확인할 수 있나요?  
-> 일반적으로 독립성은 별도의 검정 p-value로 확인하는 것이 아니라
-> 데이터 수집 구조, Feature 엔지니어링 구조를 통해 판단한다.

**Q3.** 정규성과 등분산성 가정은 각각 충족되나요?  
->필수 2 문제에서는 정규성과 등분산성 위반 근거가 없다.

**Q4.** `equal_var`에는 어떤 값을 사용해야 하나요?  
-> Levene p-value 가 0.05보다 크므로 True를 사용한다.

**Q5.** 두 집단의 평균 판매가격에는 통계적으로 유의한 차이가 있나요?
-> 유익한 차이가 있다. t검정 결과 p-value가 약 0.0015로 0.05보다 작다.
-> 일반적으로 주방설계를 잘해놓으면 주택가격이 달라질 가능성이 있다.


---

## 과제. 대응표본과 독립표본 상황 구분

### 문제 3-1. 같은 주택의 보수 전후 예상 판매가격 비교

#### 문제 설명

부동산 회사가 동일한 주택 10채에 대해 보수 전 예상 판매가격과 보수 후 예상 판매가격을 각각 산정했습니다. 같은 위치의 값은 동일한 주택의 전후 가격으로 서로 짝을 이룹니다.

```python
before_price = [145, 162, 178, 155, 190, 172, 168, 181, 159, 175]
after_price  = [154, 170, 185, 164, 201, 179, 176, 190, 168, 183]
```

단위는 천 달러입니다.

#### 요구사항

1. 두 배열의 길이가 같은지 확인하세요.
2. 보수 전후 평균을 계산하세요.
3. 독립표본 t검정과 대응표본 t검정 중 적절한 방법을 선택하고 이유를 설명하세요.
4. 대응표본 t검정의 정규성 가정은 개별 배열이 아니라 `after_price - before_price` 차이값에 적용된다는 점을 확인하세요.
5. 차이값에 Shapiro-Wilk 정규성 검정을 수행하세요.
6. 적절한 t검정을 실행하고 t통계량과 p-value를 출력하세요.
7. 보수 전후 평균 예상 판매가격에 유의한 차이가 있는지 결론을 작성하세요.

#### 해석 질문

**Q1.** 이 데이터는 독립표본인가요, 대응표본인가요?  
**Q2.** 대응표본 t검정에서는 무엇의 정규성을 확인해야 하나요?  
**Q3.** `stats.ttest_ind()`가 아니라 어떤 함수를 사용해야 하나요?  
**Q4.** 검정 결과 보수 전후 예상 판매가격에 유의한 차이가 있나요?

#### 제출 결과

- 표본 관계 판단과 근거
- 전후 평균과 차이값
- 차이값의 정규성 결과
- 대응표본 t검정 결과
- 결과 해석
- Q1~Q4 답변

In [4]:
# 과제 코드를 작성하세요.
alpha = 0.05

before_price = np.array([145, 162, 178, 155, 190, 172, 168, 181, 159, 175])
after_price  = np.array([154, 170, 185, 164, 201, 179, 176, 190, 168, 183])

# 1) 두 배열의 길이 확인
print(f"보수 전 n={len(before_price)}, 보수 후 n={len(after_price)}")
print("길이 일치 여부:", len(before_price) == len(after_price))

# 2) 전후 평균과 차이값
diff_price = after_price - before_price

print(f"\n보수 전 평균: {before_price.mean():,.2f}천달러")
print(f"보수 후 평균: {after_price.mean():,.2f}천달러")
print(f"평균 차이(후-전): {diff_price.mean():,.2f}천달러")
print("차이값:", diff_price)

# 3) 표본 관계 판단
print("\n표본 관계: 동일한 주택 10채의 보수 전후 가격이므로 대응표본")

# 4) 차이값의 정규성 검정 (개별 배열이 아니라 차이값에 적용)
shapiro_diff = stats.shapiro(diff_price)
normal_diff = shapiro_diff.pvalue > alpha

print(f"\n차이값 정규성 p-value: {shapiro_diff.pvalue:.4f} → "
      f"{'정규성 위반 근거 없음' if normal_diff else '정규성 위반 근거 있음'}")

# H0: 보수 전후 평균 예상 판매가격은 같다.
# H1: 보수 전후 평균 예상 판매가격은 다르다.

# 5) 대응표본 t검정
t_result = stats.ttest_rel(after_price, before_price)

print(f"\nt통계량: {t_result.statistic:,.4f}")
print(f"p-value: {t_result.pvalue:.4e}")
print("최종 판단:", "보수 전후 평균에 유의한 차이가 있음"
      if t_result.pvalue <= alpha else "보수 전후 평균에 유의한 차이가 없음")

보수 전 n=10, 보수 후 n=10
길이 일치 여부: True

보수 전 평균: 168.50천달러
보수 후 평균: 177.00천달러
평균 차이(후-전): 8.50천달러
차이값: [ 9  8  7  9 11  7  8  9  9  8]

표본 관계: 동일한 주택 10채의 보수 전후 가격이므로 대응표본

차이값 정규성 p-value: 0.1574 → 정규성 위반 근거 없음

t통계량: 22.8079
p-value: 2.8420e-09
최종 판단: 보수 전후 평균에 유의한 차이가 있음


### 과제 답변 작성란

**Q1.** 이 데이터는 독립표본인가요, 대응표본인가요?  
->  대응표본, 같은 대상을 두번 측정한 것으로 동일한 주택의 보수 전/후 가격이라 관측값이 1:1로 짝을 이루고 있다.

**Q2.** 대응표본 t검정에서는 무엇의 정규성을 확인해야 하나요?  
-> 차이값의 정규성. 대응표본 t검정은 차이값의 평균이 0인가를 묻는 단일표본 검정이라 가정도 차이값에 적용된다. Shapiro p-value 차이값이 0.1574로 0.05보다 커서 정규성 위반 근거가 없다.

**Q3.** `stats.ttest_ind()`가 아니라 어떤 함수를 사용해야 하나요?  
-> stats.ttest_rel(). 같은 대상을 두 번 측정해 짝이 있는 데이터를 유지한채 대응표본 t검정 함수를 사용한다. 
-> ttest_ind()는 두 배열을 서로 무관한 집단으로 보고 계산해서 짝 정보가 사라진다.

**Q4.** 검정 결과 보수 전후 예상 판매가격에 유의한 차이가 있나요?
->유의한 차이가 있음. p-value가 2.8420e-09(= 0.0000000028)로 0.05보다 작아서 보수 전후 평균이 같다는 귀무가설을 기각한다. 그리고 보수 후 예상 판매가격이 평균 8500달러 높아서 유의미한 차이가 있다고 볼 수 있다.


---

## 실습 마무리

1. t통계량은 어떤 두 값을 비교하여 계산하나요?
-> 관측된 평균 차이를 표준오차(그 차이의 불확실성)와 비교한다.

2. 단일표본 t검정은 어떤 상황에서 사용하나요?
-> 하나의 표본평균을 특정 기준값 또는 모집단 평균과 비교할 때 사용한다.

3. 독립표본과 대응표본은 데이터 수집 구조에서 어떤 차이가 있나요?
-> 독립표본은 서로 다른 대상들로 구성되고, 대응표본은 같은 대상의 반복 측정값처럼 각 관측값이 짝을 이룬다.

4. 독립표본 t검정에서 정규성·등분산성·독립성은 각각 어떻게 확인하나요?
-> Shapiro, Levene, 독립성은 데이터 수집 구조

5. 분산이 같다고 보기 어렵거나 확신할 수 없을 때 어떤 t검정을 사용할 수 있나요?
-> equal_var = False인 Welch t 검정을 사용할 수 있다.
